File: next/Journal/Boot.fs
namespace Wanxiangshu.Next.Journal

open System
open Fable.Core
open Fable.Core.JsInterop
open Wanxiangshu.Next.Kernel.Identity

module private NodeFsBoot =
    [<Import("existsSync", "node:fs")>]
    let existsSync (path: string) : bool = jsNative

    [<Import("readdirSync", "node:fs")>]
    let readdirSync (path: string) : string array = jsNative

    [<Import("openSync", "node:fs")>]
    let openSync (path: string, flags: string) : int = jsNative

    [<Import("readSync", "node:fs")>]
    let readSync (fd: int, buffer: obj, offset: int, length: int, position: obj) : int = jsNative

    [<Import("closeSync", "node:fs")>]
    let closeSync (fd: int) : unit = jsNative

    [<Import("statSync", "node:fs")>]
    let statSync (path: string) : obj = jsNative

    [<Import("join", "node:path")>]
    let pathJoin (a: string, b: string) : string = jsNative

    [<Import("basename", "node:path")>]
    let pathBasename (p: string) : string = jsNative

type Frontier = Map<RuntimeId, int64>

type BootSnapshot =
    { Envelopes: Envelope list
      Frontier: Frontier
      Diagnostics: string list }

module Boot =

    let private getRuntimeIdFromFilename (filePath: string) : RuntimeId =
        let name = NodeFsBoot.pathBasename filePath
        let idx = name.LastIndexOf('.')
        let cleanName = if idx > 0 then name.Substring(0, idx) else name
        RuntimeId.create cleanName

    let private getStatSize (stat: obj) : int64 =
        stat?size |> unbox<double> |> int64

    let captureFrontiers (directory: string) : Frontier =
        if not (NodeFsBoot.existsSync directory) then
            Map.empty
        else
            NodeFsBoot.readdirSync directory
            |> Array.filter (fun filePath -> filePath.EndsWith(".ndjson"))
            |> Array.map (fun name ->
                let filePath = NodeFsBoot.pathJoin (directory, name)
                let runtimeId = getRuntimeIdFromFilename filePath
                let stat = NodeFsBoot.statSync filePath
                let size = getStatSize stat
                (runtimeId, size))
            |> Map.ofArray

    let private readPrefixEnvelopes (filePath: string) (frontierBytes: int64) : Envelope list * string list =
        if not (NodeFsBoot.existsSync filePath) then
            [], []
        else
            let stat = NodeFsBoot.statSync filePath
            let actualFileSize = getStatSize stat
            let readLen = min frontierBytes actualFileSize
            if readLen <= 0L then
                [], []
            else
                let fd = NodeFsBoot.openSync (filePath, "r")
                let mutable res = [], []
                try
                    let buffer = Array.zeroCreate<byte> (int readLen)
                    let bytesRead = NodeFsBoot.readSync (fd, buffer, 0, int readLen, null)
                    let effectiveBytes =
                        if bytesRead <= 0 then [||]
                        else
                            let mutable lastNewline = -1
                            let mutable i = bytesRead - 1
                            while i >= 0 && lastNewline = -1 do
                                if buffer.[i] = 10uy then
                                    lastNewline <- i
                                i <- i - 1

                            if lastNewline = -1 then
                                [||]
                            else
                                buffer.[0 .. lastNewline]

                    let text = System.Text.Encoding.UTF8.GetString(effectiveBytes)
                    let lines = text.Split([| "\r\n"; "\n" |], StringSplitOptions.RemoveEmptyEntries)
                    let expectedRuntimeId = getRuntimeIdFromFilename filePath

                    let rec collect idx expectedSeq acc =
                        if idx >= lines.Length then
                            List.rev acc, []
                        else
                            match Envelope.deserialize lines.[idx] with
                            | Ok env ->
                                if env.RuntimeId <> expectedRuntimeId then
                                    let diag = sprintf "RuntimeId mismatch in %s: expected %s, got %s" (NodeFsBoot.pathBasename filePath) (RuntimeId.value expectedRuntimeId) (RuntimeId.value env.RuntimeId)
                                    List.rev acc, [ diag ]
                                else
                                    let seqVal = LocalSeq.value env.LocalSeq
                                    if seqVal <> expectedSeq then
                                        let diag = sprintf "LocalSeq anomaly in %s: expected %d, got %d" (NodeFsBoot.pathBasename filePath) expectedSeq seqVal
                                        List.rev acc, [ diag ]
                                    else
                                        collect (idx + 1) (expectedSeq + 1L) (env :: acc)
                            | Error err ->
                                let diag =
                                    sprintf "Failed to parse line %d in %s: %s" idx (NodeFsBoot.pathBasename filePath) err
                                List.rev acc, [ diag ]

                    res <- collect 0 1L []
                with ex ->
                    res <- [], [ sprintf "IO error reading %s: %s" (NodeFsBoot.pathBasename filePath) ex.Message ]
                try NodeFsBoot.closeSync fd with _ -> ()
                res

    let kWayMerge (streams: Envelope list list) : Envelope list =
        let rec merge (queues: Envelope list list) acc =
            let active = queues |> List.filter (not << List.isEmpty)

            if List.isEmpty active then
                List.rev acc
            else
                let heads = active |> List.map List.head

                let minHeadEnv =
                    heads
                    |> List.reduce (fun acc env -> if Envelope.compareSortKey env acc < 0 then env else acc)

                let rec pickAndRemove headsList =
                    match headsList with
                    | [] -> [], []
                    | (hd :: tl) :: rest when Envelope.compareSortKey hd minHeadEnv = 0 -> tl :: rest, [ hd ]
                    | q :: rest ->
                        let remainingQueues, picked = pickAndRemove rest
                        q :: remainingQueues, picked

                let nextQueues, picked = pickAndRemove active
                merge nextQueues (picked.Head :: acc)

        merge streams []

    let boot (directory: string) : BootSnapshot =
        let frontier = captureFrontiers directory

        if not (NodeFsBoot.existsSync directory) then
            { Envelopes = []
              Frontier = Map.empty
              Diagnostics = [] }
        else
            let files =
                NodeFsBoot.readdirSync directory
                |> Array.filter (fun f -> f.EndsWith(".ndjson"))

            let mutable allDiags = []

            let streamEnvelopes =
                files
                |> Array.map (fun filename ->
                    let filePath = NodeFsBoot.pathJoin (directory, filename)
                    let runtimeId = getRuntimeIdFromFilename filePath
                    let len = Map.tryFind runtimeId frontier |> Option.defaultValue 0L
                    let envs, diags = readPrefixEnvelopes filePath len
                    allDiags <- allDiags @ diags
                    envs)
                |> Array.toList

            let mergedEnvelopes = kWayMerge streamEnvelopes

            { Envelopes = mergedEnvelopes
              Frontier = frontier
              Diagnostics = allDiags }

File: .wanxiangshu/runtimes/Journal/Boot.js

import { Record } from "../fable_modules/fable-library-js.5.6.0/Types.js";
import { EnvelopeModule_compareSortKey, EnvelopeModule_deserialize, Envelope_$reflection } from "./Envelope.js";
import { record_type, string_type, class_type, int64_type, list_type } from "../fable_modules/fable-library-js.5.6.0/Reflection.js";
import { LocalSeqModule_value, RuntimeIdModule_value, RuntimeIdModule_create, RuntimeId_$reflection } from "../Kernel/Identity.js";
import { join, basename } from "node:path";
import { split, printf, toText, substring } from "../fable_modules/fable-library-js.5.6.0/String.js";
import { op_Addition, equals as equals_1, toInt32_unchecked, compare as compare_1, min, fromFloat64, toInt64_unchecked } from "../fable_modules/fable-library-js.5.6.0/BigInt.js";
import { closeSync, readSync, openSync, readdirSync, statSync, existsSync } from "node:fs";
import { tryFind, ofArray, empty } from "../fable_modules/fable-library-js.5.6.0/Map.js";
import { equals, defaultOf, compare } from "../fable_modules/fable-library-js.5.6.0/Util.js";
import { item, map } from "../fable_modules/fable-library-js.5.6.0/Array.js";
import { append, ofArray as ofArray_1, tail, head, map as map_1, reduce, isEmpty, filter, cons, reverse, singleton, empty as empty_1 } from "../fable_modules/fable-library-js.5.6.0/List.js";
import { get_UTF8 } from "../fable_modules/fable-library-js.5.6.0/Encoding.js";
import { defaultArg } from "../fable_modules/fable-library-js.5.6.0/Option.js";

export class BootSnapshot extends Record {
    constructor(Envelopes, Frontier, Diagnostics) {
        super();
        this.Envelopes = Envelopes;
        this.Frontier = Frontier;
        this.Diagnostics = Diagnostics;
    }
}

export function BootSnapshot_$reflection() {
    return record_type("Wanxiangshu.Next.Journal.BootSnapshot", [], BootSnapshot, () => [["Envelopes", list_type(Envelope_$reflection())], ["Frontier", class_type("Microsoft.FSharp.Collections.FSharpMap`2", [RuntimeId_$reflection(), int64_type])], ["Diagnostics", list_type(string_type)]]);
}

function Boot_getRuntimeIdFromFilename(filePath) {
    const name = basename(filePath);
    const idx = name.lastIndexOf(".") | 0;
    return RuntimeIdModule_create((idx > 0) ? substring(name, 0, idx) : name);
}

function Boot_getStatSize(stat) {
    return toInt64_unchecked(fromFloat64(stat.size));
}

export function Boot_captureFrontiers(directory) {
    let array;
    if (!existsSync(directory)) {
        return empty({
            Compare: (x, y) => (compare(x, y) | 0),
        });
    }
    else {
        return ofArray(map((name) => {
            const filePath_1 = join(directory, name);
            return [Boot_getRuntimeIdFromFilename(filePath_1), Boot_getStatSize(statSync(filePath_1))];
        }, (array = readdirSync(directory), array.filter((filePath) => filePath.endsWith(".ndjson")))), {
            Compare: (x_1, y_1) => (compare(x_1, y_1) | 0),
        });
    }
}

function Boot_readPrefixEnvelopes(filePath, frontierBytes) {
    let arg_9, arg_10;
    if (!existsSync(filePath)) {
        return [empty_1(), empty_1()];
    }
    else {
        const readLen = min(frontierBytes, Boot_getStatSize(statSync(filePath)));
        if (compare_1(readLen, 0n) <= 0) {
            return [empty_1(), empty_1()];
        }
        else {
            const fd = openSync(filePath, "r") | 0;
            let res = [empty_1(), empty_1()];
            try {
                const buffer = new Uint8Array(~~toInt32_unchecked(readLen));
                const bytesRead = readSync(fd, buffer, 0, ~~toInt32_unchecked(readLen), defaultOf()) | 0;
                let effectiveBytes;
                if (bytesRead <= 0) {
                    effectiveBytes = (new Uint8Array([]));
                }
                else {
                    let lastNewline = -1;
                    let i = bytesRead - 1;
                    while ((i >= 0) && (lastNewline === -1)) {
                        if (item(i, buffer) === 10) {
                            lastNewline = (i | 0);
                        }
                        i = ((i - 1) | 0);
                    }
                    effectiveBytes = ((lastNewline === -1) ? (new Uint8Array([])) : buffer.slice(0, lastNewline + 1));
                }
                const lines = split(get_UTF8().getString(effectiveBytes), ["\r\n", "\n"], undefined, 1);
                const expectedRuntimeId = Boot_getRuntimeIdFromFilename(filePath);
                const collect = (idx_mut, expectedSeq_mut, acc_mut) => {
                    collect:
                    while (true) {
                        const idx = idx_mut, expectedSeq = expectedSeq_mut, acc = acc_mut;
                        if (idx >= lines.length) {
                            return [reverse(acc), empty_1()];
                        }
                        else {
                            const matchValue = EnvelopeModule_deserialize(item(idx, lines));
                            if (matchValue.tag === 1) {
                                let diag_2;
                                const arg_7 = basename(filePath);
                                diag_2 = toText(printf("Failed to parse line %d in %s: %s"))(idx)(arg_7)(matchValue.fields[0]);
                                return [reverse(acc), singleton(diag_2)];
                            }
                            else {
                                const env = matchValue.fields[0];
                                if (!equals(env.RuntimeId, expectedRuntimeId)) {
                                    let diag;
                                    const arg = basename(filePath);
                                    const arg_1 = RuntimeIdModule_value(expectedRuntimeId);
                                    const arg_2 = RuntimeIdModule_value(env.RuntimeId);
                                    diag = toText(printf("RuntimeId mismatch in %s: expected %s, got %s"))(arg)(arg_1)(arg_2);
                                    return [reverse(acc), singleton(diag)];
                                }
                                else {
                                    const seqVal = LocalSeqModule_value(env.LocalSeq);
                                    if (!equals_1(seqVal, expectedSeq)) {
                                        let diag_1;
                                        const arg_3 = basename(filePath);
                                        diag_1 = toText(printf("LocalSeq anomaly in %s: expected %d, got %d"))(arg_3)(expectedSeq)(seqVal);
                                        return [reverse(acc), singleton(diag_1)];
                                    }
                                    else {
                                        idx_mut = (idx + 1);
                                        expectedSeq_mut = toInt64_unchecked(op_Addition(expectedSeq, 1n));
                                        acc_mut = cons(env, acc);
                                        continue collect;
                                    }
                                }
                            }
                        }
                        break;
                    }
                };
                res = collect(0, 1n, empty_1());
            }
            catch (ex) {
                res = [empty_1(), singleton((arg_9 = basename(filePath), (arg_10 = ex.message, toText(printf("IO error reading %s: %s"))(arg_9)(arg_10))))];
            }
            try {
                closeSync(fd);
            }
            catch (matchValue_1) {
            }
            return res;
        }
    }
}

export function Boot_kWayMerge(streams) {
    const merge = (queues_mut, acc_mut) => {
        merge:
        while (true) {
            const queues = queues_mut, acc = acc_mut;
            const active = filter((arg) => !isEmpty(arg), queues);
            if (isEmpty(active)) {
                return reverse(acc);
            }
            else {
                const minHeadEnv = reduce((acc_1, env) => {
                    if (EnvelopeModule_compareSortKey(env, acc_1) < 0) {
                        return env;
                    }
                    else {
                        return acc_1;
                    }
                }, map_1(head, active));
                const pickAndRemove = (headsList) => {
                    let matchResult, hd_1, rest_1, tl_1, q, rest_2;
                    if (!isEmpty(headsList)) {
                        if (!isEmpty(head(headsList))) {
                            if (EnvelopeModule_compareSortKey(head(head(headsList)), minHeadEnv) === 0) {
                                matchResult = 1;
                                hd_1 = head(head(headsList));
                                rest_1 = tail(headsList);
                                tl_1 = tail(head(headsList));
                            }
                            else {
                                matchResult = 2;
                                q = head(headsList);
                                rest_2 = tail(headsList);
                            }
                        }
                        else {
                            matchResult = 2;
                            q = head(headsList);
                            rest_2 = tail(headsList);
                        }
                    }
                    else {
                        matchResult = 0;
                    }
                    switch (matchResult) {
                        case 0:
                            return [empty_1(), empty_1()];
                        case 1:
                            return [cons(tl_1, rest_1), singleton(hd_1)];
                        default: {
                            const patternInput = pickAndRemove(rest_2);
                            return [cons(q, patternInput[0]), patternInput[1]];
                        }
                    }
                };
                const patternInput_1 = pickAndRemove(active);
                queues_mut = patternInput_1[0];
                acc_mut = cons(head(patternInput_1[1]), acc);
                continue merge;
            }
            break;
        }
    };
    return merge(streams, empty_1());
}

export function Boot_boot(directory) {
    let array;
    const frontier = Boot_captureFrontiers(directory);
    if (!existsSync(directory)) {
        return new BootSnapshot(empty_1(), empty({
            Compare: (x, y) => (compare(x, y) | 0),
        }), empty_1());
    }
    else {
        let allDiags = empty_1();
        return new BootSnapshot(Boot_kWayMerge(ofArray_1(map((filename) => {
            const filePath = join(directory, filename);
            const patternInput = Boot_readPrefixEnvelopes(filePath, defaultArg(tryFind(Boot_getRuntimeIdFromFilename(filePath), frontier), 0n));
            allDiags = append(allDiags, patternInput[1]);
            return patternInput[0];
        }, (array = readdirSync(directory), array.filter((f) => f.endsWith(".ndjson")))))), frontier, allDiags);
    }
}

